### PR TITLE
[GEOS-11382] The interceptor "CiteComplianceHack" never gets invoked by the Dispatcher Servlet

### PR DESCRIPTION
**Pull request by afabiani**
[GEOS-11382](https://osgeo-org.atlassian.net/browse/GEOS-11382) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

Fixes: https://osgeo-org.atlassian.net/browse/GEOS-11382
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/wms/src/main/java/applicationContext.xml [hunk 1]
```diff
@@ -408,10 +414,12 @@
...
-    <bean id="wmsCiteComplianceHack" class="org.geoserver.ows.CiteComplianceHack">
-        <constructor-arg ref="geoServer"/>
-        <constructor-arg value="org.geoserver.wms.WMSInfo"/>
-    </bean>
+    <mvc:interceptors>
+        <bean id="wmsCiteComplianceHack" class="org.geoserver.ows.CiteComplianceHack">
+            <constructor-arg ref="geoServer"/>
+            <constructor-arg value="org.geoserver.wms.WMSInfo"/>
+        </bean>
+    </mvc:interceptors>
```
Comment (aaime) [inline src/wms/src/main/java/applicationContext.xml:422]: I heard that in order to make WMS compliant, one had to make all the other services compliant... but this change seems to be related to having the interceptor be actually invoked.  The same interceptor is declared for other services as well, why has it been changed only here?

I've done a quick test and here are some thoughts.

The situation is confusing. Each OGC service should have its own compliance hack declaration, but only a few actually do, WMS, WFS, and CSW (but not WCS for example, maybe it's strict by default):

```xml
C:\Users\andre\devel\geoserver>git grep -i citecompliancehack -- "*.xml"
src/extension/csw/core/src/main/java/applicationContext.xml:          <ref bean="cswCiteComplianceHack"/>
src/extension/csw/core/src/main/java/applicationContext.xml:    <bean id="cswCiteComplianceHack" class="org.geoserver.ows.CiteComplianceHack">
src/wfs/src/main/java/applicationContext.xml:   <bean id="wfsCiteComplianceHack" class="org.geoserver.ows.CiteComplianceHack">
src/wfs/src/main/java/applicationContext.xml:                           <ref bean="wfsCiteComplianceHack"/>
src/wms/src/main/java/applicationContext.xml:    <bean id="wmsCiteComplianceHack" class="org.geoserver.ows.CiteComplianceHack">
```

Originally the all used to run, maybe some Spring upgrade broke that? In any case, looking at the compliance hack source code, it makes no attempt to check which service is being called. So if all 3 of the above hacks are called, the last one wins, which is a random choice picking the cite compliance configuration of the last hack that gets run.

The hack should be doing some attempt at figuring out if the service is a match, e.g., check the last element of the path, and/or check if there is a service=<id> kvp parameter that matches the current service configuration.

My other worry is that the mvc configuration is something new, and it might end up interacting with the other dispatchers, like the REST API and the OGC API one... hopefully not, the cite compliance hack specifically checks that the handler is the OWS Dispatcher. But I'd recommend doing some light manual testing of the REST API and the OGC API in a running GeoServer.

Long story short:
* Update the hack to check its own configured service only
* Update other OGC services with CITE hacks as well
* Add a test to ensure the flag is working, it's evidently missing or we would not have had this regression

Comment (afabiani) [reply src/wms/src/main/java/applicationContext.xml:422]: Yessir!

Comment (afabiani) [reply src/wms/src/main/java/applicationContext.xml]: @aaime turns out that there's no need to introduce the `mvc:interceptors` tag. There was just a missing initialized element on the `wmsURLMapping` bean.

I've been added few more tests and checks for the current service.

File: src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java [hunk 2]
```diff
@@ -36,12 +44,66 @@ public boolean preHandle(
...
+        // create a new request instance
+        Request req = new Request();
+
+        // set request / response
+        req.setHttpRequest(request);
+        req.setHttpResponse(response);
+
+        // find the service
+        try {
+            return dispatcher.service(req);
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java:69]: Good idea, but problematic. The service method at the end calls the dispatcher callbacks, which should not happen:
* they would receive events out of order
* they would receive events multiple times (when the dispatcher actually parses the request)

That said, it should be easy to refactour a method that only computes the service descriptor, make it static, and call it from both the existing Dispatcher.service(...) and from here.

I'm worried about POST requests, as the XML body is read while computing the service... it is properly "pushed back" so that the dispatcher can parse it again? You might want to test a GetFeature in XML format, interactively and/or as part of tests.

File: src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java [hunk 3]
```diff
@@ -36,12 +44,66 @@ public boolean preHandle(
...
+        } catch (Throwable t) {
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java:70]: Never catch a Throable, you want Error instances (e..g, OOM) to go free. Catch Exception instead.

File: src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java [hunk 4]
```diff
@@ -36,12 +44,66 @@ public boolean preHandle(
...
+                        req.setServiceDescriptor(serviceDescriptor);
+                        try {
+                            return dispatcher.fireServiceDispatchedCallback(req, serviceDescriptor);
+                        } catch (ServiceException se) {
+                            LOGGER.log(
+                                    Level.FINE,
+                                    "Exception while searching for the 'Service Descriptor'",
+                                    se);
+                            return null;
+                        }
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java:94]: Why go through the callbacks, and not just return the service found?

File: src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java [hunk 5]
```diff
@@ -36,12 +44,66 @@ public boolean preHandle(
...
+        // set request / response
+        req.setHttpRequest(request);
+        req.setHttpResponse(response);
+
+        // find the service
+        try {
+            return dispatcher.service(req);
+        } catch (Throwable t) {
+            LOGGER.log(Level.FINE, "Exception while looking for the 'Service' from the request", t);
+            // load from teh context
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java:72]:
```suggestion
            // load from the context
```

File: src/wfs/src/main/java/applicationContext.xml [hunk 6]
```diff
@@ -1,11 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- 
+<!--
  Copyright (C) 2014 - 2016 Open Source Geospatial Foundation. All rights reserved.
  This code is licensed under the GPL 2.0 license, available at the root
  application directory.
  -->
 <!DOCTYPE beans PUBLIC "-//SPRING//DTD BEAN//EN" "http://www.springframework.org/dtd/spring-beans.dtd">
-
```
Comment (aaime) [inline src/wfs/src/main/java/applicationContext.xml:8]: Several unrelated changes in this file (whitespace)

File: src/wms/src/main/java/applicationContext.xml [hunk 7]
```diff
@@ -1,12 +1,12 @@
-
-<!--
```
Comment (aaime) [inline src/wms/src/main/java/applicationContext.xml:2]: More whiltespace unrelated changes

File: src/wms/src/main/java/applicationContext.xml [hunk 8]
```diff
@@ -418,6 +418,11 @@
 		class="org.geoserver.ows.OWSHandlerMapping">
 		<constructor-arg ref="catalog"/>
 		<property name="alwaysUseFullPath" value="true"/>
+        <property name="interceptors">
```
Comment (aaime) [inline src/wms/src/main/java/applicationContext.xml:421]: This block is mis-aligned:

image

**Review by aaime — CHANGES_REQUESTED** [review]
This is definitely going in the right direction. See comments below.

**Comment by aaime** [comment]
Looks good to me. Pay attention the QA build is failing.

File: src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java (hunk 3 above)
Comment (afabiani) [reply src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java]: done

File: src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java (hunk 4 above)
Comment (afabiani) [reply src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java]: done

File: src/wfs/src/main/java/applicationContext.xml (hunk 6 above)
Comment (afabiani) [reply src/wfs/src/main/java/applicationContext.xml]: done

File: src/wms/src/main/java/applicationContext.xml (hunk 8 above)
Comment (afabiani) [reply src/wms/src/main/java/applicationContext.xml]: done

**Review by aaime — APPROVED** [review]
